File: lua/conversor.lua
local function paraBinario(numero)
    if numero == 0 then return "0" end
    local binario = ""
    while numero > 0 do
        binario = tostring(numero % 2) .. binario
        numero = numero // 2
    end
    return binario
end

local function paraOctal(numero)
    if numero == 0 then return "0" end
    local octal = ""
    while numero > 0 do
        octal = tostring(numero % 8) .. octal
        numero = numero // 8
    end
    return octal
end

local function main(arg)
    if #arg ~= 1 then
        print("Uso: lua5.4 conversor.lua <hexadecimal>")
        os.exit(1)
    end

    local entrada = arg[1]
    local valorHex = entrada:gsub("^0[xX]", "")

    local numeroDecimal = tonumber(valorHex, 16)

    if not numeroDecimal then
        print("Erro: '" .. entrada .. "' não é um hexadecimal válido.")
        os.exit(1)
    end

    print("Hexadecimal: " .. entrada:upper())
    print("Decimal:     " .. numeroDecimal)
    print("Binário:     " .. paraBinario(numeroDecimal))
    print("Octal:       " .. paraOctal(numeroDecimal))
end

main(arg)

File: csharp/Conversor.cs
using System;

class Conversor
{
    static string ParaBinario(long numero)
    {
        if (numero == 0) return "0";
        return Convert.ToString(numero, 2);
    }

    static string ParaOctal(long numero)
    {
        if (numero == 0) return "0";
        return Convert.ToString(numero, 8);
    }

    static void Main(string[] args)
    {
        Console.OutputEncoding = new System.Text.UTF8Encoding(false);

        if (args.Length != 1)
        {
            Console.WriteLine("Uso: Conversor.exe <hexadecimal>");
            Environment.Exit(1);
        }

        string entrada = args[0];
        string valorHex = entrada.Trim();
        if (valorHex.StartsWith("0x", StringComparison.OrdinalIgnoreCase))
        {
            valorHex = valorHex.Substring(2);
        }

        long numeroDecimal;
        try
        {
            numeroDecimal = Convert.ToInt64(valorHex, 16);
        }
        catch (FormatException)
        {
            Console.WriteLine($"Erro: '{entrada}' não é um hexadecimal válido.");
            Environment.Exit(1);
            return;
        }

        Console.WriteLine($"Hexadecimal: {entrada.ToUpper()}");
        Console.WriteLine($"Decimal:     {numeroDecimal}");
        Console.WriteLine($"Binário:     {ParaBinario(numeroDecimal)}");
        Console.WriteLine($"Octal:       {ParaOctal(numeroDecimal)}");
    }
}
